sdk-base/src/main/java/com/mapbox/maps/plugin/viewport/experimental/state/ViewportState.kt

```diff
@@ -0,0 +1,28 @@
+package com.mapbox.maps.plugin.viewport.experimental.state
+
+import com.mapbox.maps.plugin.viewport.ViewportPlugin
+import com.mapbox.maps.plugin.viewport.experimental.data.ViewportDataSource
+import com.mapbox.maps.plugin.viewport.experimental.transition.ViewportStateTransition
+
+/**
+ * Set of possible [ViewportPlugin] states.
+ */
+interface ViewportState {
+  val id: String
+  val isTransitioning: Boolean
+  val dataSource: ViewportDataSource?
+  val stateTransition: ViewportStateTransition?
+
```

Comment:
Let's limit this interface to only what's necessary for the viewport to use a state.

* dataSource — does the viewport need to do anything with this? why not let it be an implementation detail of the state?
* stateTransition — in a model where you have 3 states, each state could need a transition to each other state. I think we need a different way of representing that. I would have the viewport own the transitions instead of the states.
* id — is it necessary?

Seems like there should be some other interfaces here like something the viewport can use to tell a state when it becomes current or is no longer current

Reply:
> dataSource — does the viewport need to do anything with this? why not let it be an implementation detail of the state?

Just tried to put in the minimum following state implementation, and figured out that we don't really need to expose the dataSource, but we do need a way to configure the `ViewportDataSourceOptions`, which is a part of the `ViewportDataSource` class. I have removed it from the public interface in the latest iteration.

> stateTransition — in a model where you have 3 states, each state could need a transition to each other state. I think we need a different way of representing that. I would have the viewport own the transitions instead of the states.

The state should own the transition so that users can customise how to transition from previous state to the current state. However, this could be a implementation detail and we could just add the following function to the public interface of the state.
```
fun transitionToState(
    stateTransitionOptions: ViewportTransitionOptions,
    frameTransitionOptions: ViewportTransitionOptions,
    transitionEndListener: TransitionEndListener?
  )
```

> id — is it necessary?
> 

This is necessary as we could allow user to use our customised Following and overview state implementation, with the pre-defined id. Users could also `addState` once in the beginning and reuse the customised state with the ID and avoid keeping a reference of the customised state themselves.


@macdrevx I have updated the PR to have a minimum follow state implementation and hooked it up to the test app, please review again.

Reply:
Updated the public interfaces as we discussed offline, please take a look.